pkg/backend/httpstate/backend.go

```diff
@@ -1129,7 +1131,64 @@ func (b *cloudBackend) Preview(ctx context.Context, stack backend.Stack,
 func (b *cloudBackend) Update(ctx context.Context, stack backend.Stack,
 	op backend.UpdateOperation,
 ) (sdkDisplay.ResourceChanges, error) {
-	return backend.PreviewThenPromptThenExecute(ctx, apitype.UpdateUpdate, stack, op, b.apply)
+	return backend.PreviewThenPromptThenExecute(ctx, apitype.UpdateUpdate, stack, op, b.apply, b)
+}
+
+func (b *cloudBackend) isCopilotFeatureEnabled(projectName tokens.PackageName, opts display.Options) bool {
+	// Have copilot features been requested by specifying the --copilot flag to the cli
+	if !opts.ShowCopilotFeatures {
+		return false
+	}
+
+	// Is copilot enabled this project in Pulumi Cloud
+	enabled, ok := b.copilotEnabledForProject[projectName]
+	contract.Assertf(ok,
+		"copilotEnabledForProject has not been set for project %q. only available after an update has been started.",
+		projectName)
+	return enabled
+}
+
+// explain takes engine events, renders them out to a buffer as something similar to what the user sees
+// in the CLI, and then explains the output with Copilot.
+func (b *cloudBackend) explain(
+	stackRef backend.StackReference,
+	op backend.UpdateOperation,
+	events []engine.Event,
+	opts display.Options,
+) (string, error) {
+	renderer := display.NewCaptureProgressEvents(
+		stackRef.Name(),
+		op.Proj.Name,
+		display.Options{},
+		true,
+		apitype.UpdateUpdate,
+	)
+	renderer.ProcessEventSlice(events)
+	output := renderer.Output()
```

Comment:
So we are copying exactly the way the preview renders these events, and we render all event types? Don't we want more control over what exactly we send to the API?

Looking at the implementation of the `details` option, it has a custom loop iterating over the slice of events, some filtering of events, some re-ordering of events, etc. Keeping in mind the other discussion that we may need to render the details of events for AI to look at, will it mean we'll have to change the rendering here?

Reply:
> Keeping in mind the other discussion that we may need to render the details of events for AI to look at, will it mean we'll have to change the rendering here?

Yes we've found that Summary + Details is giving the best explanations right now. (and is now implemented here via https://github.com/pulumi/pulumi/pull/19138/commits/2005437b40e64658e6a6ce082ffb87e4d65854bb).

Having the rendering done server side would allow us to continue tweaking this in the future for older clients too.. hmm.

Reply:
But yes using the ProgressEvents renderer on this event set with `ShowResourceChanges: true` gives the same output as the base summary lines + details.